File: 9/HasPrefix.go
package main

import "fmt"

// HasPrefix is just to solve 計算機科学一問一答
func HasPrefix(s, prefix string) bool {
	return len(s) >= len(prefix) && s[:len(prefix)] == prefix
}

// HasSuffix is just to solve 計算機科学一問一答
func HasSuffix(s, suffix string) bool {
	return len(s) >= len(suffix) && s[len(s)-len(suffix):] == suffix
}

// Contains is just to solve 計算機科学一問一答
func Contains(s, substr string) bool {
	for i := 0; i < len(s); i++ {
		if HasPrefix(s[i:], substr) {
			return true
		}
	}
	return false
}

func main() {
	s := "Hello, world"
	hello := "Hello"
	world := "world"

	fmt.Println("HasPrefix(), expecting: true, false |", HasPrefix(s, hello), HasPrefix(s, world))
	fmt.Println("HasSuffix(), expecting: false, true |", HasSuffix(s, hello), HasSuffix(s, world))
	fmt.Println("Contains(), expecting: true, true, false |", Contains(s, hello), Contains(s, world), Contains(s, "something"))
}

File: 9/has-prefix.py
#!/usr/bin/env python3


def has_prefix(s, prefix):
    return len(s) >= len(prefix) and s[:len(prefix)] == prefix


def has_suffix(s, suffix):
    print(s[len(suffix):])
    return len(s) >= len(suffix) and s[len(s)-len(suffix):] == suffix


def contains(s, substr):
    for i in range(len(s) - len(substr) + 1):
        if has_prefix(s[i:], substr):
            return True
    return False

if __name__ == "__main__":
    s = "Hello, world"
    hello = "Hello"
    world = "world"

    print("expecting: True, False |", has_prefix(s, hello), has_prefix(s, world))
    print("expecting: False, True |", has_suffix(s, hello), has_suffix(s, world))
    print("expecting: true, true, false |", contains(s, hello), contains(s, world), contains(s, "something"))
